src/requirements/Allowlist/components/AllowlistFormInputs.tsx

```diff
@@ -48,103 +53,194 @@ export default function AllowlistFormInputs({
   const router = useRouter()
 
   const isHidden = useWatch({ name: `${baseFieldPath}.data.hideAllowlist` })
+  const requirementFileId = useWatch({ name: `${baseFieldPath}.data.fileId` })
+  const addressCount = useWatch({ name: `${baseFieldPath}.data.addressCount` })
+  const status = useWatch({ name: `${baseFieldPath}.data.status` })
 
-  const parseFile = (file: File) => {
-    const fileReader = new FileReader()
+  const requirementType = useWatch({ name: `${baseFieldPath}.type` })
 
-    fileReader.onload = () => {
-      const lines = fileReader.result
-        ?.toString()
-        ?.split("\n")
-        ?.filter((line) => !!line)
-        ?.map((line) => line.trim())
+  const uploadFileToGcs = useSubmit(async (file: File) => {
+    const [data, body] = await Promise.all([
+      fetcher(
+        `${process.env.NEXT_PUBLIC_API.replace(
+          "/v1",
+          "/v2"
+        )}/third-party/google/signed-url?requirementType=${
+          requirementType || "ALLOWLIST"
+        }`
+      ),
+      file.arrayBuffer(),
+    ])
 
-      setValue(`${baseFieldPath}.data.addresses`, lines, {
-        shouldValidate: true,
-        shouldDirty: true,
-      })
+    const { signedUrl, fileId } = data || {}
+
+    if (!signedUrl || !fileId) {
+      throw new Error("Failed to upload allowlist")
     }
 
-    fileReader.readAsText(file)
-  }
+    const gcsResponse = await fetch(signedUrl, {
+      method: "PUT",
+      body,
+      headers: {
+        "Content-Type": "text/csv",
+      },
+    })
+
+    if (!gcsResponse.ok) {
+      throw new Error("Failed to upload allowlist")
+    }
+
+    setValue(`${baseFieldPath}.data.addresses`, [], {
+      shouldDirty: true,
+    })
+
+    setValue(`${baseFieldPath}.data.fileId`, fileId, {
+      shouldDirty: true,
+    })
+
+    return fileId
+  })
 
   const { isDragActive, fileRejections, getRootProps, getInputProps } = useDropzone({
     multiple: false,
-    accept: { "text/*": [".csv", ".txt"] },
+    accept: { "text/*": [".csv"] },
     onDrop: (accepted) => {
-      if (accepted.length > 0) parseFile(accepted[0])
+      if (accepted.length > 0) {
+        uploadFileToGcs.onSubmit(accepted[0])
+      }
     },
+    maxSizeMb: 50,
   })
 
   const type = useWatch({ name: `${baseFieldPath}.type` })
 
   return (
     <Stack spacing={4} alignItems="start" {...getRootProps()}>
       <FormControl isInvalid={!!fileRejections?.[0]} textAlign="left">
-        <FormLabel>Upload from file</FormLabel>
+        <FormLabel display={"flex"} gap={2}>
+          Upload from file
+          {addressCount && (
+            <Tooltip
+              label={`Currently an address with ${addressCount} addresses is uploaded. Uploading a new file overwrites these addresses`}
```

Comment:
I think this should be "Currently an allowlist...", hehe? 👀

Reply:
Nice catch 😄 
https://github.com/guildxyz/guild.xyz/pull/1244/commits/1e66494587c36314e664d0c9d4246f1ac0e2af9b